schema_enforcer/instances/file.py

```diff
@@ -62,53 +71,97 @@ def __init__(self, root, filename, matches=None):
         """
         self.data = None
         self.path = root
         self.full_path = os.path.realpath(root)
         self.filename = filename
 
+        # Internal vars for caching data
+        self._top_level_properties = set()
+
         if matches:
             self.matches = matches
         else:
-            self.matches = []
+            self.matches = set()
+
+        self._add_matches_by_decorator()
+
+    @property
+    def top_level_properties(self):
+        """Return a list of top level properties in the structured data defined by the data pulled from _get_content.
+
+        Returns:
+            top_level_properties (set): Set of the strings of top level properties defined by the data file
+        """
+        if self._top_level_properties:
+            return self._top_level_properties
 
-        self.matches.extend(self._find_matches_inline())
+        content = self._get_content()
+        top_level_properties = set(content.keys())
+        return top_level_properties
 
-    def _find_matches_inline(self, content=None):
-        """Find addition matches using the Schema ID decorator comment.
+    def _add_matches_by_decorator(self, content=None):
+        """Add matches which declare schema IDs they should adhere to using a decorator comment.
 
-        Look for a line with # jsonschema: schema_id,schema_id
+        If a line of the form # jsonschema: <schema_id>,<schema_id> is defined in the data file, the
+        schema IDs will be added to the list of schema IDs the data will be checked for adherence to.
 
         Args:
             content (string, optional): Content of the file to analyze. Default to None.
 
         Returns:
-            list(string): List of matches found in the file.
+            set(string): Set of matches (strings of schema_ids) found in the file.
         """
         if not content:
-            content = Path(os.path.join(self.full_path, self.filename)).read_text()
+            content = self._get_content(structured=False)
 
-        matches = []
+        matches = set()
 
         if SCHEMA_TAG in content:
             line_regexp = r"^#.*{0}:\s*(.*)$".format(SCHEMA_TAG)
             match = re.match(line_regexp, content, re.MULTILINE)
             if match:
-                matches = [x.strip() for x in match.group(1).split(",")]
+                matches = {x.strip() for x in match.group(1).split(",")}
 
-        return matches
+        matches = matches - self.matches
+        self.matches.update(matches)
 
-    def get_content(self):
-        """Return the content of the instance file in structured format.
+    def _get_content(self, structured=True):
+        """Return the content of the instance file.
 
         Content returned can be either dict or list depending on the content of the file
 
+        Args:
+            structured (bool): Whether to return structured or unstructured data. Defaultes to true.
+
         Returns:
             dict or list: Content of the instance file.
```

Comment:
What's the return type if `structured==False`? Also might need to update line 134 above.

Reply:
Updated to denote behavior.